src/core/bpf-lsm.c

```diff
@@ -0,0 +1,264 @@
+        if (inner_map_fd < 0)
+                return log_debug_errno(errno, "Failed to create BPF map");
+
+        r = bpf_object_set_inner_map_fd(obj, "cgroup_hash", inner_map_fd);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set inner map fd: %m");
+
+        r = bpf_object_load(obj);
+        if (r < 0)
+                return log_error_errno(errno, "Failed to load LSM BPF program: %m");
+
+        /* Close dummy map */
+        close(inner_map_fd);
+
+        r = bpf_object_find_program_by_title(obj, LSM_PROG_SECTION_NAME, &prog);
+        if (r < 0) {
+                return log_error_errno(r, "Failed to find LSM BPF program '%s'", LSM_PROG_SECTION_NAME);
+        }
+
+        link = bpf_program__attach_lsm(prog);
+        if (!link)
+                return log_error_errno(errno, "Failed to attach BPF LSM program: %m");
+
+        log_info("LSM BPF program attached");
+
+        outer_map_fd = bpf_object_get_map_fd(obj, "cgroup_hash");
+        if (outer_map_fd < 0)
+                return log_error_errno(outer_map_fd, "Failed to get map fd: %m");
+
+        r = mkdir_parents(LSM_MAP_BPFFS_PATH, 0700);
+        if (r < 0)
+                return log_error_errno(r, "Failed create systemd bpffs directory: %m");
+
+        r = bpf_obj_pin(outer_map_fd, LSM_MAP_BPFFS_PATH);
+        if (r < 0)
+                return log_error_errno(r, "Failed to pin LSM BPF map in the bpffs: %m");
+
+        return 0;
+}
+
+int bpf_restrict_filesystems(char **filesystems, char *cgroup_path) {
+        _cleanup_close_ int inner_map_fd = -1, outer_map_fd = -1;
+        _cleanup_free_ char *path = NULL;
+        uint64_t cgroup_id;
+        uint32_t magic, dummy_value = 1;
+        char **f;
+        int r;
+
+        assert(filesystems);
+        assert(cgroup_path);
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, cgroup_path, NULL, &path);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get systemd cgroup path");
+
+        r = cg_path_get_cgroupid(path, &cgroup_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get cgroup ID for path '%s'", path);
+
+        inner_map_fd = bpf_create_map(
+            BPF_MAP_TYPE_HASH,
+            sizeof(uint32_t),
+            sizeof(uint32_t),
+            128, /* Should be enough for all filesystem types */
+            0);
+        if (inner_map_fd < 0)
+                return log_error_errno(errno, "Failed to create inner LSM map: %m");
+
+        outer_map_fd = bpf_obj_get(LSM_MAP_BPFFS_PATH);
+        if (outer_map_fd < 0)
+                return log_error_errno(errno, "Error getting pinned LSM BPF map: %m");
+
+        if (bpf_map_update_elem(outer_map_fd, &cgroup_id, &inner_map_fd, BPF_ANY) != 0)
+                return log_error_errno(errno, "Error populating LSM BPF map: %m");
+
+        STRV_FOREACH(f, filesystems) {
+                if (DEBUG_LOGGING)
```

Comment:
drop this conditionalization, it's implicit in `log_debug()`.

(you want to use `if (DEBUG_LOGGING)` if you need to do some expensive work to be able to print the debug message, but this is not the case here, hence just call log_debug() directly)